### ISSUE TITLE
Old packages in Ubunutu 12.04 cause apache conf breakage

### ISSUE THREAD

**Issue by pde** [letsencrypt/letsencrypt#2012]
Despite landing #1961, some people have continued to report parser failures with it...

**Comment by ehsaan** [comment]
Take a look. Let me know if you want me to try out a change/...: https://goo.gl/bIE9T4

**Comment by rroseselavy42** [comment]
can also confirm

**Comment by cimm** [comment]
Same problem here. Tried Let's Encrypt a few weeks back and ran into issue #1961. Tried it again today with the latest master (ffd0d5a) but still failing with the `PluginError('Unable to parse runtime variables',)` error. Running Apache/2.2.22 on Debian 7.

Not running `mod_define` if that's important. Here are the loaded modules:

```
 core_module (static)
 log_config_module (static)
 logio_module (static)
 version_module (static)
 mpm_prefork_module (static)
 http_module (static)
 so_module (static)
 alias_module (shared)
 auth_basic_module (shared)
 authn_file_module (shared)
 authz_default_module (shared)
 authz_groupfile_module (shared)
 authz_host_module (shared)
 authz_user_module (shared)
 autoindex_module (shared)
 cgi_module (shared)
 deflate_module (shared)
 dir_module (shared)
 env_module (shared)
 headers_module (shared)
 mime_module (shared)
 negotiation_module (shared)
 php5_module (shared)
 reqtimeout_module (shared)
 rewrite_module (shared)
 setenvif_module (shared)
 ssl_module (shared)
 status_module (shared)
```

**Comment by grantnhoward** [comment]
Also confirmed, still unresolved with latest master.
Running 2.2.22 on Ubuntu 12.04
ran ```./letsencrypt-auto --apache ```
received:

```
The apache plugin is not working; there may be problems with your existing configuration.
The error was: PluginError('Unable to parse runtime variables',)
```

letsencrypt.log
```
2015-12-27 22:06:57,200:DEBUG:letsencrypt.cli:letsencrypt version: 0.1.1
2015-12-27 22:06:57,200:DEBUG:letsencrypt.cli:Arguments: ['--apache', '-v']
2015-12-27 22:06:57,200:DEBUG:letsencrypt.cli:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2015-12-27 22:06:57,204:DEBUG:letsencrypt.cli:Requested authenticator apache and installer apache
2015-12-27 22:06:57,261:DEBUG:letsencrypt.plugins.disco:Other error:(PluginEntryPoint#apache): Unable to parse runtime variables
Traceback (most recent call last):
  File "/home/user/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/plugins/disco.py", line 103, in prepare
    self._initialized.prepare()
  File "/home/user/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/configurator.py", line 148, in prepare
    self.aug, self.conf("server-root"), self.conf("ctl"))
  File "/home/user/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 39, in __init__
    self.update_runtime_variables(ctl)
  File "/home/user/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 104, in update_runtime_variables
    raise errors.PluginError("Unable to parse runtime variables")
PluginError: Unable to parse runtime variables
2015-12-27 22:06:57,262:DEBUG:letsencrypt.display.ops:No candidate plugin
2015-12-27 22:06:57,262:DEBUG:letsencrypt.cli:Selected authenticator None and installer None
```

Maybe this is of some use.

**Comment by floh1111** [comment]
Same problem here on Debian Wheezy using Apache/2.2.22.

Installed Modules (mod_define is not installed)
```
root@XXXXXX:~/letsencrypt# apache2ctl -M
Loaded Modules:
 core_module (static)
 log_config_module (static)
 logio_module (static)
 version_module (static)
 mpm_worker_module (static)
 http_module (static)
 so_module (static)
 actions_module (shared)
 alias_module (shared)
 auth_basic_module (shared)
 authn_file_module (shared)
 authz_default_module (shared)
 authz_groupfile_module (shared)
 authz_host_module (shared)
 authz_user_module (shared)
 autoindex_module (shared)
 cgi_module (shared)
 cgid_module (shared)
 deflate_module (shared)
 dir_module (shared)
 env_module (shared)
 fastcgi_module (shared)
 mime_module (shared)
 negotiation_module (shared)
 reqtimeout_module (shared)
 rewrite_module (shared)
 setenvif_module (shared)
 ssl_module (shared)
 status_module (shared)
 suexec_module (shared)
Syntax OK
```

Letsencrypt log:
```
2015-12-28 08:53:42,437:DEBUG:letsencrypt.cli:Root logging level set at 30
2015-12-28 08:53:42,437:INFO:letsencrypt.cli:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2015-12-28 08:53:42,437:DEBUG:letsencrypt.cli:letsencrypt version: 0.1.1
2015-12-28 08:53:42,437:DEBUG:letsencrypt.cli:Arguments: ['--apache']
2015-12-28 08:53:42,438:DEBUG:letsencrypt.cli:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2015-12-28 08:53:42,442:DEBUG:letsencrypt.cli:Requested authenticator apache and installer apache
2015-12-28 08:53:42,509:DEBUG:letsencrypt.plugins.disco:Other error:(PluginEntryPoint#apache): Unable to parse runtime variables
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/plugins/disco.py", line 103, in prepare
    self._initialized.prepare()
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/configurator.py", line 148, in prepare
    self.aug, self.conf("server-root"), self.conf("ctl"))
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 39, in __init__
    self.update_runtime_variables(ctl)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 104, in update_runtime_variables
    raise errors.PluginError("Unable to parse runtime variables")
PluginError: Unable to parse runtime variables
2015-12-28 08:53:42,510:DEBUG:letsencrypt.display.ops:No candidate plugin
2015-12-28 08:53:42,510:DEBUG:letsencrypt.cli:Selected authenticator None and installer None
```

**Comment by joohoi** [comment]
@floh1111 @grantnhoward You both seem to be running a release version (0.1.1) which doesn't have the changes. Consult documentation on how to run the current (dev) version, which has the PR #1961

**Comment by ixevix** [comment]
@joohoi which documentation would that be?

I tried this but the log still says I'm running 0.1.1

https://letsencrypt.readthedocs.org/en/latest/contributing.html#running-a-local-copy-of-the-client

**Comment by ixevix** [comment]
Ok now I found at the end of that documentation: 

Debian
sudo ./bootstrap/debian.sh

that gives:
No libaugeas0 version is available that's new enough to run the
Let's Encrypt apache plugin...

**Comment by ixevix** [comment]
Same error after installing libaugeas0 from backports

**Comment by joohoi** [comment]
You created the venv and executed letsencrypt (not letsencrypt-auto) with the bootstrapped venv activated?

**Comment by ixevix** [comment]
Ok now it broke my whole apache config. how do I rollback?

**Comment by ixevix** [comment]
Ok it seems it disabled mod_mime which resulted in problems with coldfusion. Now the https version does not work however.

**Comment by ixevix** [comment]
I needed to enable some mods that letsencrypt? probably disabled: ssl and setenvif. Are these changes documented somewhere or do we have to guess what letsencrypt does to our apache configs?

**Comment by tking** [comment]
For me it still does not work:
2016-01-03 19:14:26,140:DEBUG:letsencrypt.cli:Root logging level set at 20
2016-01-03 19:14:26,141:INFO:letsencrypt.cli:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2016-01-03 19:14:26,185:DEBUG:letsencrypt.cli:letsencrypt version: 0.1.1
2016-01-03 19:14:26,185:DEBUG:letsencrypt.cli:Arguments: ['--apache', '-d', 't-king.de', '-d', 'www.t-king.de', '-v']
2016-01-03 19:14:26,185:DEBUG:letsencrypt.cli:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2016-01-03 19:14:26,198:DEBUG:letsencrypt.cli:Requested authenticator apache and installer apache
2016-01-03 19:14:26,310:DEBUG:letsencrypt.plugins.disco:Other error:(PluginEntryPoint#apache): Unable to parse runtime variables
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/plugins/disco.py", line 103, in prepare
    self._initialized.prepare()
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/configurator.py", line 148, in prepare
    self.aug, self.conf("server-root"), self.conf("ctl"))
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 39, in __init__
    self.update_runtime_variables(ctl)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 104, in update_runtime_variables
    raise errors.PluginError("Unable to parse runtime variables")
PluginError: Unable to parse runtime variables
2016-01-03 19:14:26,313:DEBUG:letsencrypt.display.ops:No candidate plugin
2016-01-03 19:14:26,314:DEBUG:letsencrypt.cli:Selected authenticator None and installer None

Can I provide any help/support debugging the situation?

**Comment by SwartzCr** [comment]
@tking it looks like ou're running version 0.1.1 which is doesn't have these changes, can you follow these instructions and try again with this release branch: https://letsencrypt.readthedocs.org/en/latest/contributing.html#running-a-local-copy-of-the-client

**Comment by SwartzCr** [comment]
@ixevix can you confirm that the problem is that Let's Encrypt is failing to carry over certain mod statements into your new config file? such as mod_mime?
Would you be willing to share a (sanitized) version of your config that I can use for testing?

**Comment by ixevix** [comment]
@SwartzCr I'll get back to you with more information later. Time for sleep now.

**Comment by ixevix** [comment]
At least these are failing:

`AddHandler jakarta-servlet .cfm .cfml .cfc .cfr .cfswf`
from
`/etc/apache2/mod_jk.conf`
and
`        <IfModule mod_php5.c>
                AddType application/x-httpd-php .php`
from
`/etc/apache2/conf.d/phpmyadmin.conf`

when you disable mod_mime

**Comment by ixevix** [comment]
the mod_jk is a custom module for coldfusion

`LoadModule    jk_module  "/opt/coldfusion10/config/wsconfig/1/mod_jk.so"`

**Comment by SwartzCr** [comment]
@cimm did you run letsencrypt-auto or did you make a venv to ensure that you are running the code that you downloaded? You can look here for specific instructions on running a local copy of the client:
https://letsencrypt.readthedocs.org/en/latest/contributing.html#running-a-local-copy-of-the-client

**Comment by SwartzCr** [comment]
@ixevix : just to clarify - Let's Encrypt is completing and claiming success, but the generated conf file is lacking a mod_mime statement?

**Comment by ixevix** [comment]
no the letsencrypt run fails with errors

`2015-12-28 14:18:43,812:DEBUG:letsencrypt.cli:Root logging level set at 30
2015-12-28 14:18:43,812:INFO:letsencrypt.cli:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2015-12-28 14:18:43,812:DEBUG:letsencrypt.cli:letsencrypt version: 0.2.0.dev0
2015-12-28 14:18:43,812:DEBUG:letsencrypt.cli:Arguments: []
2015-12-28 14:18:43,813:DEBUG:letsencrypt.cli:Discovered plugins: PluginsRegistry(PluginEntryPoint#nginx,PluginEntryPoint#standalone,PluginEntryPoint#manual,PluginEntryPoint#webroot,PluginEntryPoint#apache,PluginEntryPoint#null)
2015-12-28 14:18:43,849:ERROR:letsencrypt.le_util:Error while running apache2ctl configtest.
Action 'configtest' failed.
The Apache error log may have more information.

Syntax error on line 14 of /etc/apache2/mod_jk.conf:
Invalid command 'AddHandler', perhaps misspelled or defined by a module not included in the server configuration

2015-12-28 14:18:43,857:DEBUG:letsencrypt.plugins.disco:Misconfigured PluginEntryPoint#apache: Error while running apache2ctl configtest.
Action 'configtest' failed.
The Apache error log may have more information.

Syntax error on line 14 of /etc/apache2/mod_jk.conf:
Invalid command 'AddHandler', perhaps misspelled or defined by a module not included in the server configuration
Traceback (most recent call last):
  File "/home/ixevix/tmp/letsencrypt/letsencrypt/plugins/disco.py", line 103, in prepare
    self._initialized.prepare()
  File "/home/ixevix/tmp/letsencrypt/letsencrypt-apache/letsencrypt_apache/configurator.py", line 157, in prepare
    self.config_test()
  File "/home/ixevix/tmp/letsencrypt/letsencrypt-apache/letsencrypt_apache/configurator.py", line 1331, in config_test
    raise errors.MisconfigurationError(str(err))
MisconfigurationError: Error while running apache2ctl configtest.
Action 'configtest' failed.
The Apache error log may have more information.

Syntax error on line 14 of /etc/apache2/mod_jk.conf:
Invalid command 'AddHandler', perhaps misspelled or defined by a module not included in the server configuration

2015-12-28 14:18:43,859:DEBUG:letsencrypt.plugins.disco:No installation (PluginEntryPoint#nginx):
Traceback (most recent call last):
  File "/home/ixevix/tmp/letsencrypt/letsencrypt/plugins/disco.py", line 103, in prepare
    self._initialized.prepare()
  File "/home/ixevix/tmp/letsencrypt/letsencrypt-nginx/letsencrypt_nginx/configurator.py", line 112, in prepare
    raise errors.NoInstallationError
NoInstallationError
2015-12-28 14:18:43,859:DEBUG:letsencrypt.display.ops:Single candidate plugin: * apache
Description: Apache Web Server - Alpha
Interfaces: IAuthenticator, IInstaller, IPlugin
Entry point: apache = letsencrypt_apache.configurator:ApacheConfigurator
Initialized: <letsencrypt_apache.configurator.ApacheConfigurator object at 0x27cf310>
Prep: Error while running apache2ctl configtest.
Action 'configtest' failed.
The Apache error log may have more information.

Syntax error on line 14 of /etc/apache2/mod_jk.conf:
Invalid command 'AddHandler', perhaps misspelled or defined by a module not included in the server configuration`

**Comment by ixevix** [comment]
My guess is letsencrypt disables the required modules before running apache2ctl configtest which then fails.

**Comment by joohoi** [comment]
The error trace suggests that your apache configuration is broken to begin with. You don't get any errors by running apache2ctl configtest manually? 

"/home/ixevix/tmp/letsencrypt/letsencrypt-apache/letsencrypt_apache/configurator.py", line 157 
configtest is ran before the client does anything to your configuration. It's there to check that your config is sane.

**Comment by ixevix** [comment]
I'm guessing something happened before that so that the configuration was broken. I have a rather long log file that I would like to share. It seems I have ran letsencrypt with the following arguments: ['--installer', 'apache']. Is there a way to share it privately?

**Comment by ixevix** [comment]
@joohoi I sent the logs by email.

**Comment by ixevix** [comment]
And yes. I got the same errors by running it manually.

**Comment by joohoi** [comment]
LE can't fix your broken apache config for you. I'd suggest fixing it manually and making sure apache2ctl passes, after that try running LE again.

**Comment by ixevix** [comment]
I'm reluctant to run letsencrypt again since the config was fine before running it. Did you receive the log files?

**Comment by joohoi** [comment]
I did, the only file that has any signifigance is the .4 log file that shows LE enabling modules in your env. It will also deactivate them in case of error, with the rollback functionality (and this is the only case it disables anything for the user)

**Comment by ixevix** [comment]
Ok, I remember running rollback after it failed. That's where it must have gone wrong.

**Comment by joohoi** [comment]
It seems to have enabled ssl module, mime module and setenvif module. One of these caused apache to parse aome sections of your existing configuration, that failed. It seems to point to the same file and same line as the later errors. It pretty much did wake up the syntax error beast sleeping in your config files, prolly inside some ifmodule - section. However it's impossible to tell without the actual configuration files.

**Comment by joohoi** [comment]
So my suggestion still applies. Try fixing the configuration manually, and rerun LE after. Also pull the latest master with latest patches.

**Comment by ixevix** [comment]
I don't know what I should fix. It keeps telling me I've got No vhost selected. Maybe there is a problem with running several sites?

**Comment by ixevix** [comment]
Should the config files be okay with ssl, mime and setenvif disabled? Because that is not the case.

**Comment by ixevix** [comment]
Also every time I run letsencrypt it disables ssl, mime and setenvif so I have to manually enable them and restart apache so I can get my sites back up.

**Comment by ixevix** [comment]
So i fix configtest, run letsencrypt and run configtest again directly after and it fails.

**Comment by ixevix** [comment]
Do you suggest grepping for something specific in /etc/apache2 ?

**Comment by joohoi** [comment]
The configfiles should be ok with ssl, mime and setenvif enabled. Could you paste the /var/log/letsencrypt/letsencrypt.log for the run after fixing the configuration manually to get a clearer view what's going on.

**Comment by ixevix** [comment]
Now it cries:

`2016-01-05 21:39:03,018:ERROR:letsencrypt_apache.configurator:No vhost exists with servername or alias of: xxx. No vhost was selected. Please specify servernames in the Apache config`

**Comment by joohoi** [comment]
Can you post your apache configuration? It sounds like configuration issue, but it's impossible to debug this further without the files.

**Comment by ixevix** [comment]
Maybe I can do that. Should I just tar up the whole /etc/apache2 folder? or just the sites-enabled files or what?

**Comment by ixevix** [comment]
Ubuntu 12.04 has some packages that are too old: https://github.com/letsencrypt/letsencrypt/pull/1811

**Comment by joohoi** [comment]
It wasn't apache 2.2 related, but distribution related.

**Comment by SwartzCr** [comment]
@ehsaan and @rroseselavy42 can you confirm that your problems aren't also from package issues? Can you provide your OS versions?

**Comment by SwartzCr** [comment]
@ixevix can you list which packages are out of date? We can solve updating them like we did with libaugeas0: https://github.com/letsencrypt/letsencrypt/pull/1961/files#diff-a91a7c93bee71eb52c3a7d15354a0394R38

**Comment by ixevix** [comment]
@SwartzCr actually I was just told by @joohoi to post it. He might have a better idea.

**Comment by joohoi** [comment]
Yup, I don't have a clear view which packages are out of date either. Just that the exactly same configuration fails to recognize any vhosts on Ubuntu 12.04 running Apache2.2 where as Debian wheezy running Apache 2.2 works beautifully. Also bunch of tests fail on Ubuntu 12.04, but I didn't have time to debug it further.

**Comment by pde** [comment]
@ixevix yes please post a tarball of your `/etc/apache2/` if there's nothing secret in there.
@joohoi are you saying you can reproduce on Ubuntu 12 but not Debian 7?  Or just that you can't reproduce on Debian 7?

**Comment by ixevix** [comment]
@pde the problem is not in the configuration, it's in the Ubuntu packages.

**Comment by joohoi** [comment]
@pde I recieved @ixevix configuration tarball, and we fixed the configuration itself. After fixing it, it worked beautifully in Apache 2.2 @ Debian 7 and LE found the vhosts. However when transfering the same exact configuration to Apache 2.2 @ Ubuntu 12.04 LE failed to find any vhosts.

On Ubuntu 12.04 even the default LE tests fail, you can test it out quick by just changing our root Vagrantfile to load ubuntu/precise64 instead of ubuntu/trusty64 image.

**Comment by pde** [comment]
Hmmm.  Older Ubuntu 12.04 systems might not have a sufficiently new libaugeas0, which could cause that problem.  But the newer version was initially added to a ppa, and then backported.  But maybe some users aren't getting it?  @ixevix what version of libaugeas0 do you have installed?

**Comment by ixevix** [comment]
I tried with `0.10.0-0ubuntu4` from the stable dist. No dice.

**Comment by pde** [comment]
@ixevix you'll need to enable precise-backports to get the newer version.  We'll need to update our bootstrap scripts to do this automatically on Ubuntu 12.04 system where it isn't the default situation.  The Apache plugin should also check the augeas version before it declares itself available, I think...

**Comment by pde** [comment]
Let's close this ticket since the title is now about bootstrapping issues that have been fixed in 0.2.0.  However we will need another one for Apache 2.2 actually working in 0.2.1...
